src/Frontend/Core.hs

```diff
@@ -150,9 +167,20 @@ instance Aeson.ToJSON (PostResult a) where toJSON = Aeson.gtoJson
 -- Using this via `curl` is complicated by the fact that we need cookie authentication, so this
 -- feature should be used via the 'createPageSamples' mechanism (see "Frontend" and 'footerMarkup'
 -- for more details).
-type GetH = Get '[HTML, PlainText]
-type PostH p = Post '[HTML] (PostResult p)
-type FormHandler p = FormH '[HTML, PlainText] (Frame (FormPageRep p)) (FormPageResult p)
+type GetH p = Get '[HTML, PlainText] (GetResult p)
+type PostH' p r = Post '[HTML, PlainText] (PostResult p r)
+type PostH p = PostH' p ()
+type FormHandler p =
+       GetH (Frame (FormPageRep p))
+  :<|> FormReqBody :> PostH' (Frame (FormPageRep p)) (Frame (FormPageRep p)) -- Redirect
```

Comment:
Redirect looks like a good fit for the second parameter of `PostH'` (which could replace `PostH` if we add `()` at a few places). However this does not work because of validation errors, in case of errors we do not redirect but display `Frame (FormPageRep p)`.